src/lem/circuit.rs

```diff
@@ -0,0 +1,889 @@
+//! construct an **implication system**, which is responsible for ensuring that
+//! allocated variables in the concrete path are equal to their expected values
+//! but such equalities on virtual paths are irrelevant.
+//!
+//! ## Slot optimization:
+//!
+//! Some operations like Poseidon hash can be relatively expensive in the circuit,
+//! therefore we want to avoid wasting constraints with them as much as possible.
+//! In order to achieve this goal, we only add constraints to a limited number of
+//! *slots* that correspond to a maximum number of such expensive operations that
+//! may occur in a single interpretation round. This optimization avoids
+//! constraining every expensive operation on every possible path.
+//!
+//! Shortly, we construct the hash slot system using the next steps:
+//!
+//! * STEP 1: Static analysis, when we traverse LEM for the first time and allocate
+//! slots for hashes in all virtual paths. We only add hashes that were not yet
+//! previously added, therefore ensuring deduplication of hashes in different
+//! branches.
+//!
+//! * STEP 2: During interpretation (second traversal) we gather information
+//! related to each visited slot.
+//!
+//! * STEP 3: During construction of constraints, we first preallocate the preimage
+//! and image for each slot; then we synthesize all slots; and finally we traverse
+//! LEM (for the third time), adding implication constraints, glueing together all the
+//! information. This step is better explained in the `synthesize` function.
+
+use std::collections::{HashMap, HashSet};
+
+use anyhow::{anyhow, bail, Context, Result};
+use bellperson::{
+    gadgets::{boolean::Boolean, num::AllocatedNum},
+    ConstraintSystem,
+};
+use indexmap::{IndexMap, IndexSet};
+
+use crate::circuit::gadgets::{
+    constraints::{
+        alloc_equal_const, and, enforce_selector_with_premise, implies_equal, implies_equal_zero,
+    },
+    data::hash_poseidon,
+    pointer::AllocatedPtr,
+};
+
+use crate::field::{FWrap, LurkField};
+
+use super::{
+    interpreter::{Frame, SlotType},
+    path::Path,
+    pointers::ZPtr,
+    store::Store,
+    MetaPtr, LEM, LEMOP,
+};
+
+/// Holds a counter per path. We want to use this to count the number of slots
+/// that have already been used on each LEM path.
+#[derive(Default)]
+struct PathTicker(HashMap<Path, usize>);
+
+impl PathTicker {
+    /// Increments the counter of a path. If the path wasn't tracked, returns
+    /// `0` and starts tracking it
+    pub(crate) fn tick(&mut self, path: Path) -> usize {
+        let next = self.0.get(&path).unwrap_or(&0).to_owned();
+        self.0.insert(path, next + 1);
+        next
+    }
+
+    /// Starts tracking a new path with a counter from another. If the reference
+    /// path wasn't being tracked, the new one won't be either, such that calling
+    /// `next` will return `0`.
+    pub(crate) fn cont(&mut self, new: Path, from: &Path) {
```

Comment:
Is it required for correctness that `from` always be a prefix of `new`? If so, can we assert that here, please?

Better, can we enforce it by constructing the new path *here*?

Best: can paths alsot not be *strings*?

Reply:
Will do